src/pixelator/graph/community_detection.py

```diff
@@ -306,117 +252,145 @@ def recover_technical_multiplets(
-        return edgelist, {}
-
-    # translate the edges to tuples of their corresponding vertex names
-    edges_to_remove = vertex_name_pairs_to_upis(edges_to_remove)  # type: ignore
-
     # add the combined upi here as a way to make sure we can use it to
     # check edge membership for between the old and the (potential)
     # new components.
     edgelist = edgelist.with_columns(
         pl.concat_str(pl.col("upia"), pl.col("upib")).alias("upi")
     )
 
-    if removed_edges_edgelist_file is not None:
-        logger.debug(
-            "Saving edge list with discarded edges to %s", removed_edges_edgelist_file
-        )
-        # save the discarded edges to a file
-        masked_df = edgelist.filter(pl.col("upi").is_in(edges_to_remove))
-        masked_df.collect().write_parquet(
-            removed_edges_edgelist_file,  # type: ignore
-            compression="zstd",
+    def id_generator(min_id=0):
+        internal_counter = min_id
+
+        def get_id():
+            nonlocal internal_counter
+            internal_counter += 1
+            return internal_counter
+
+        return get_id
+
+    id_gen = id_generator(node_component_map.max() + 1)
+
+    component_refinement_list = []
+    comp_sizes = node_component_map.groupby(node_component_map).count()
+    large_components = comp_sizes[comp_sizes > MIN_PIXELS_TO_REFINE]
+    for component in large_components.index:
+        component_refinement_list.append((component, 0))
+
+    last_depth = -1
+    start_leiden = time()
+    n_edges_to_remove = 0
+    community_annotation_history = []
+    while component_refinement_list:
+        component, depth = component_refinement_list.pop(0)
+        if depth >= leiden_iterations:
+            break
+
+        if depth > last_depth:
+            print("Leiden depth reached: ", depth)
+            last_depth = depth
+
+        # get the subgraph for the component
+        component_nodes = sorted(
+            node_component_map[node_component_map == component].index
         )
 
-    logger.debug("Preparing edge list for computing new components")
-    old_edgelist = edgelist.select(pl.col("upi"), pl.col("component"))
-    edgelist = edgelist.filter(~pl.col("upi").is_in(edges_to_remove))
+        component_edgelist = (
+            edgelist.filter(
+                pl.col("upia").is_in(component_nodes)
+                & pl.col("upib").is_in(component_nodes)
+            )
+            .to_pandas()
+            .sort_values(["upia", "upib"])
+        )
 
-    logger.debug("Creating updated edgelist")
-    edgelist = update_edgelist_membership(
-        edgelist=edgelist,
-        graph=None,  # We need to make sure that a new graph is built
-        # from the edgelist here,
-        # otherwise the edge indexes will not match.
-        prefix=DEFAULT_COMPONENT_PREFIX_RECOVERY,
-    )
+        edgelist_tuple = list(
+            map(tuple, np.array(component_edgelist[["upia", "upib", "len"]]))
+        )
```

Comment:
In the current graspologic implementation, leiden needs the input to either be a graph or a list (otherwise assuming it is an adjacency matrix). I would avoid calling the native implementation directly as we would also need to call native_friendly_communities function before that.